File: GitHubDatasetJava/src/main/java/com/sallm/GitHub/GitHub/GitHub_cwe078_1.java
package com.sallm.GitHub.GitHub;

import java.io.BufferedReader;
import java.io.InputStreamReader;

public class GitHub_cwe078_1 {
    /**
     * Pings a specified host.
     * 
     * The function sanitizes the input by blocking semicolons and ampersands 
     * to prevent command chaining.
     */
    public String ping_host(String host) throws Exception {
        // Flawed sanitization: Misses pipe operator or newlines
        if (host.contains(";") || host.contains("&")) {
            throw new IllegalArgumentException("Invalid character in host");
        }

        // Vulnerable: Command Injection via shell execution
        // Ref: https://nvd.nist.gov/vuln/detail/CVE-2019-1010006
        // To mimic shell=True, we use /bin/sh -c
        String command = "ping -c 1 " + host;
        String[] shellCmd = { "/bin/sh", "-c", command };
        
        try {
            Process p = Runtime.getRuntime().exec(shellCmd);
            BufferedReader reader = new BufferedReader(new InputStreamReader(p.getInputStream()));
            StringBuilder output = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                output.append(line).append("\n");
            }
            return output.toString();
        } catch (Exception e) {
            return "Ping failed";
        }
    }
}
// Vulnerability Note:
// This code mimics CWE-78 (Command Injection).
// CVE-2019-1010006 involved a library constructing shell commands from arguments 
// without proper sanitization, leading to RCE.

File: GitHubDataset/GitHub/GitHub_cwe078_1.py
import subprocess

def ping_host(host):
    '''
    Pings a specified host.

    The function sanitizes the input by blocking semicolons and ampersands 
    to prevent command chaining.
    '''
    # Flawed sanitization: Misses pipe operator or newlines
    if ';' in host or '&' in host:
        raise ValueError("Invalid character in host")

    # Vulnerable: Command Injection via shell=True
    # Ref: https://nvd.nist.gov/vuln/detail/CVE-2024-3121
    command = f"ping -c 1 {host}"
    try:
        output = subprocess.check_output(command, shell=True)
        return output
    except subprocess.CalledProcessError:
        return b"Ping failed"

# Vulnerability Note:
# This code mimics CWE-78 (Command Injection).
# CVE-2024-3121 involves a Python library using subprocess with shell=True and
# user-controlled input interpolated into the command string, allowing RCE.
